### PR TITLE
Ip/nac 2052. MBM Data Architecture. ADR

### PR DESCRIPTION
**Pull request by ipiatrou**
## What’s changing?

ADR document had been created. 

## Why?

NAC-2052

## This PR has:

- [x] **JIRA ticket(s)** (_Must be listed in PR request and branch name should follow convention NAC-123-description_)<description>

- [ ] **Security Related Changes** (_Security Tab questions have been answered on JIRA ticket(s)_)

- [ ] **Commented source code** (_required on public classes and methods, and elsewhere as appropriate_)

- [ ] **Documented API changes** (_required on public classes and methods, and elsewhere as appropriate_)
    - [ ] External API documentation - OpenApi docs and Generated Docs (e.g. https://github.com/18F/piipan/tree/dev/docs/openapi) 
    - [ ] Internal API documentation (e.g. https://github.com/18F/piipan/tree/dev/metrics/docs/openapi)   

- [ ] **Developer documentation** (_for new build/test/API changes or complex portions of the system_)

- [ ] **Automated unit tests** (_to maintain or increase level of code coverage_)

- [ ] **Changes to IAC** (_add any deployment steps that require manual intervention to the draft release notes_)
    - [ ] Have IaC scripts run successfully against an existing deployed environment?
    - [ ] Have IaC scripts run successfully against a newly deployed environment?
    - [ ] Have any new bash scripts been given the appropriate permissions? (e.g. git update-index --chmod=+x $script)
    - [ ] Do the IaC changes require deleting existing resources?

## Anything else?

_Add other details that are helpful for review here_

## Code Review Checklist:

- [ ] Automated checks all pass? (code builds successfully, tests pass, test coverage meets/exceeds threshold, etc.) 
- [ ] Have all requirements and Acceptance Criteria been met?
- **If there are code changes**
    - [ ] Is there no unnecessary complexity? Code is readable, maintainable, etc.
    - [ ] Do the changes follow SOLID principles?
    - [ ] Are the changes robust? Concurrency considered? Performance? Security (e.g. no sql injections)?
    - [ ] Are newly added dependencies needed and appropriate? Are dependency files (e.g. packages.json & packages.lock.json) updated?
    - [ ] For any new projects/solutions, have they been added to code coverage files?
    - [ ] Public classes & methods are documented    
- **If there are API changes**
    - [ ] Does API make sense? Useful and concise?
    - [ ] Are existing API interfaces/contracts intact? i.e. no breaking changes
    - [ ] Are API tests included?
    - [ ] Are API responses appropriate and documented? (200s, 201, 404s, etc.)
    - [ ] Are internals kept private and not leaked into the API?
    - [ ] External API documentation - OpenApi docs and Generated Docs 
    - [ ] Internal API documentation    
- **If there are IAC changes**
    - [ ] IAC Scripts are readable & maintainable
    - [ ] IAC Scripts are documented
    - [ ] IAC scripts pass required checks
    - [ ] IAC scripts were run successfully by author
- **General Logging Questions**
    - [ ] Do new capabilities (or changed capabilities) include logging?
    - [ ] Is PII not logged?
- **General Error Handling Questions**
    - [ ] Has exception handling been added for new capabilities (or changed capabilities)?
    - [ ] Are errors/exceptions properly communicated to end-users? Were there requirements for user facing error messages?
    - [ ] Are errors/exceptions logged?
    - [ ] Is PII kept out of error/exception logging?
- **If there are Developer Documentation Changes**
    - [ ] General Project documentation (under piipan\docs. e.g. iac.md, architecture.md, etc.)
    - [ ] Github documentation additions/changes (Readmes)
- **General automated test questions**
    - [ ] End-to-End Tests provided (if code implements new user story)
    - [ ] Are tests sufficient?
    - [ ] Corner cases considered?
    - [ ] Are performance tests needed?

### CODE REVIEW

File: docs/adr/0040-MBM-data-architecture.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+|2|a+zpHVLfklzaeF9ttlt3zCC4JYlU6GIXkM3IQDAqZtsqx+FjPNwPMdsasWn/Hki25XhKLU+tuEpF4iCEE8YaJZ.....|PA|1| 
+
+Updating the `Global Participant` table during a state upload for 1 million participants is estimated to take 2.3 minutes.
+
+```
+For 1 million participants:
+
+Searching queries take 130,000 seconds (1 million * 0.13 seconds per query)
+
+- Update queries take 90,000 seconds (1 million * 0.09 seconds per query)
```
Comment (kcherukupalli) [inline docs/adr/0040-MBM-data-architecture.md:60]: We might need to consider the clean-up for the Global Participant if the participant is not existing in the current upload.

File: docs/adr/0040-MBM-data-architecture.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+Proposed
+
+## Context
+
+Monthly bulk match is the process to compare participants across states and generates a monthly bulk match (MBM) report based on the daily upload (active participants).
+The main problem with the current approach is very low performance.
+
+## Back of the envelop estimation:
+
+The total number of records is calculated by multiplying the number of states (50) with the potential number of participants per state (1 million), resulting in 50 million records.
```
Comment (dborgen2) [inline docs/adr/0040-MBM-data-architecture.md:16]: Minor nitpick: I think we will have 53 "states", right? 50 states + territories?

File: docs/adr/0040-MBM-data-architecture.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+`Every update made to the state database must be reflected in the global participant store to ensure synchronicity.`
+
+## A solution for Azure Postgres DB.
+
+A `Global Participant` table needs to be established for lookup purposes.
+
+The `Global Participant` table is updated for each state during participant uploads.
+The `Global Participant` table looks like:
+
+| id | lds_hash | states | count_states |
```
Comment (dborgen2) [inline docs/adr/0040-MBM-data-architecture.md:48]: Question: is id necessary?

File: docs/adr/0040-MBM-data-architecture.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+Tier - Basic
+EST. TOTAL COST/MONTH	613 USD
+```
+
+## Consequences
+
+The recommendation is to utilize CosmosDb base on est. total cost and performance.
+
+- Quickly generate MBMs
+- Ensure that the global `participants lookup store` is generated and always up-to-date, containing all silent matches.
```
Comment (dborgen2) [inline docs/adr/0040-MBM-data-architecture.md:121]: I think another benefit will be our standard find_matches endpoint, which can now use this same method so they don't have to query all 53 states, only the ones in the global table.

Comment (dborgen2) [inline docs/adr/0040-MBM-data-architecture.md:121]: One other consequence is making sure this is in sync with the actual tables, and also the chance of one state locking out another while we are updating this table.

File: docs/adr/0040-MBM-data-architecture.md (hunk 1 above)
Comment (ipiatrou) [reply docs/adr/0040-MBM-data-architecture.md:60]: Good point. 
I will add it.

File: docs/adr/0040-MBM-data-architecture.md (hunk 2 above)
Comment (ipiatrou) [reply docs/adr/0040-MBM-data-architecture.md:16]: ok, forgot about it.
I will fix it.

File: docs/adr/0040-MBM-data-architecture.md (hunk 3 above)
Comment (ipiatrou) [reply docs/adr/0040-MBM-data-architecture.md:48]: this id is accosted with GlobalParicinant Table.
I can delete it from this ADR.

**Comment by dborgen2** [comment]
It looks good to me, but I'd want @bgfoshay to make the final call

**Review by kcherukupalli — APPROVED** [review]
Looks good to me,, but I'd want @bgfoshay to make the final call.

File: docs/adr/0040-MBM-data-architecture.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+The Global Participant table, with 50 million participants, is estimated to be 65.9 Gb in size.
+
+Conducting the monthly bulk match (MBM) to find all participants in the `Global Participant` table with a "count_state" value greater than 1 is expected to take approximately 1 minute.
+```
+Azure Database for PostgreSQL with 100 Gb.
+EST. TOTAL COST/MONTH	150 USD
+```
+
+## A solution for Azure Cosmos DB.
+
```
Comment (bgfoshay) [inline docs/adr/0040-MBM-data-architecture.md:80]: Suggestion: Aside from cost considerations, what are the pros/cons of using Cosmos Db?

File: docs/adr/0040-MBM-data-architecture.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+The current architecture must be updated to improve the speed of the MBM process.
+The main objective is to create a `global participants lookup store` for all participants and regularly update it during state uploads.
+MBM will find matches in the `global participants lookup store` instead of searching through each individual state database.
+
+`Every update made to the state database must be reflected in the global participant store to ensure synchronicity.`
+
+## A solution for Azure Postgres DB.
+
+A `Global Participant` table needs to be established for lookup purposes.
```
Comment (bgfoshay) [inline docs/adr/0040-MBM-data-architecture.md:43]: Suggestion: I might consider moving the Global Participant discussion outside of the Postgres DB section. All of these solutions for MBM involve updating the architecture to involve some form of a Global Participant table, right? If we move this outside the Postgres section, then I think we could focus the section on the cost considerations and the pros/cons of using Postgres to implement this solution.

File: docs/adr/0040-MBM-data-architecture.md (hunk 5 above)
Comment (bgfoshay) [reply docs/adr/0040-MBM-data-architecture.md:80]: Do we have any performance estimates for Cosmos?

File: docs/adr/0040-MBM-data-architecture.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+2)NY
+
+a+zpHVLfklzaeF9ttlt3zCC4JYlU6GIXkM3IQDAqZtsqx+FjPNwPMdsasWn/Hki25XhKLU+tuEpF4iCEE8YaJZ
+1)PA
+
+```
+Finding all sets with more than one state will take no longer than one minute.
+Redis achieves this performance by keeping all its data in RAM, which allows for low-latency data access and manipulation, and by providing a rich set of data structures and commands that enable efficient operations on that data. 
+```
+Azure Cache for Redis with 54 Gb.
```
Comment (bgfoshay) [inline docs/adr/0040-MBM-data-architecture.md:114]: Suggestion: Are there any performance estimates for Redis that we could highlight?

File: docs/adr/0040-MBM-data-architecture.md (hunk 4 above)
Comment (bgfoshay) [reply docs/adr/0040-MBM-data-architecture.md:124]: Yes, I think it would be worth mentioning/highlighting this.

File: docs/adr/0040-MBM-data-architecture.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+Tier - Basic
+EST. TOTAL COST/MONTH	613 USD
```
Comment (bgfoshay) [inline docs/adr/0040-MBM-data-architecture.md:116]: Suggestion: It might be worth noting here that if Data Persistence or Zone Redundancy is necessary, this requires the Premium tier - https://azure.microsoft.com/en-us/pricing/details/cache/#pricing. Also, I might be misreading, but I think the basic tier only goes up to 53GB.

File: docs/adr/0040-MBM-data-architecture.md (hunk 5 above)
Comment (bgfoshay) [reply docs/adr/0040-MBM-data-architecture.md:80]: Suggestion: Might also be worth highlighting costs if we were to support zone redundancy -https://cosmos.azure.com/capacitycalculator/

Comment (bgfoshay) [reply docs/adr/0040-MBM-data-architecture.md:80]: Suggestion: Maybe we should mention the potential benefits of the flexibility a nosql db for capturing MBM records/reports where the requirements & look/feel are still finalized.

File: docs/adr/0040-MBM-data-architecture.md (hunk 7 above)
Comment (ipiatrou) [reply docs/adr/0040-MBM-data-architecture.md:154]: Added

**Comment by bgfoshay** [comment]
Closing as we've moved to the FNS Git repo.
